posix/unix.c

```diff
@@ -947,6 +947,46 @@ int unix_close(unsigned socket)
 }
 
 
+int unix_poll(unsigned socket, short events)
+{
+	unixsock_t *s, *conn;
+	int err = 0;
+
+	if ((s = unixsock_get(socket)) == NULL) {
+		err = POLLNVAL;
+	}
+	else {
+		if (events & (POLLIN | POLLRDNORM | POLLRDBAND)) {
+			proc_lockSet(&s->lock);
+			if (_cbuffer_avail(&s->buffer) > 0)
+				err |= events & (POLLIN | POLLRDNORM | POLLRDBAND);
+			proc_lockClear(&s->lock);
+		}
+
+		if (events & (POLLOUT | POLLRDNORM | POLLRDBAND)) {
+			if ((conn = s->connect) != NULL) {
```

Comment:
i'm wondering if we could have a race condition here (memory pointed by `conn` could be freed after this line by doing `unixsock_put` (eg. by `unixsock_close`) on the connected socket. I think that:
 - we should ensure every unixsock_put takes lock to closing socket or is called under lock
 - to modify `s->connect` in `unixsock_put` we should take `s->lock` (and every access to s->connect should be also done under the lock?)
 - (I don't know if we won't introduce some way to deadlock sockets closing, but it's done under the global lock so maybe we're safe - I don't know the usual code paths here)

Edit: maybe better would be to use `unixsock_get(s->connect_fd)` to ensure the connected socket won't be deleted while we're operating on it? (seems like a cleaner solution to me)

Reply:
Yeah, good catch. Unfortunately this code is probably full of race conditions. I don't wont to dive into this right now. In this particular case your "unixsock_get" suggestion should fix it. Another option is to acquire "s->lock" in "unixsock_put" before destroying the socket - this should also fix the same problem in "send" function.

Reply:
Yeah, IMHO we should always use `unixsock_get` when we're operating on a peer socket, but apart from the poll it's out of scope of this PR (maybe we should add a task not to forget it?). Socket deletion should also be reviewed. I was thinking about switching field `connected` to `connected_fd`, but this is not enough to fix deletion use case (and would result in strange behavior if somebody would close the connected socket and open new one reusing the FD :)).

I'll go for using `unixsock_get` here and at least adding FIXME in places we suspect race condition may occur.

Reply:
IMHO there still could be a race condition (`s->connect` could be freed between  `s->connect != NULL` access and  && `conn = unixsock_get(s->connect->id)` - maybe add FIXME there? But now the window for race condition is much smaller :)